### PR TITLE
Add a button to respond to the ExceptionWitness

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Finalized Popover
Popover closed

### CODE REVIEW

File: ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import static org.testfx.api.FxAssert.verifyThat;
+
+public class ExceptionWitnessResponderButtonTest extends ApplicationTest {
+
+    private Object witnessed;
+    private EventBus eventBus;
+    private ExceptionWitness witness;
+
+    @Override
+    public void start(Stage stage) throws Exception {
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java]: :+1:

File: ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.testfx.api.FxAssert.verifyThat;
+
+public class ExceptionWitnessResponderButtonTest extends ApplicationTest {
+
+    private Object witnessed;
+    private EventBus eventBus;
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java]: :+1:

File: ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    @Test
+    public void testOnExceptionClearedEvent() throws Exception {
+        flagNewException();
+        witness.clearException();
+        WaitForAsyncUtils.waitFor(2, TimeUnit.SECONDS,
+                () -> NodeMatchers.isInvisible().matches(lookup("." + ExceptionWitnessResponderButton.STYLE_CLASS).queryFirst()));
+    }
+
+    @Test
+    public void testPopoverAppears() throws TimeoutException {
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java]: :+1:

File: ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import org.testfx.util.WaitForAsyncUtils;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.testfx.api.FxAssert.verifyThat;
+
+public class ExceptionWitnessResponderButtonTest extends ApplicationTest {
+
+    private Object witnessed;
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java]: :+1:

File: ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Test
+    public void testOnExceptionEvent() throws Exception {
+        flagNewException();
+        WaitForAsyncUtils.waitFor(2, TimeUnit.SECONDS,
+                () -> NodeMatchers.isVisible().matches(lookup("." + ExceptionWitnessResponderButton.STYLE_CLASS).queryFirst()));
+    }
+
+    @Test
+    public void testOnExceptionClearedEvent() throws Exception {
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java]: :+1:

File: ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        eventBus = new EventBus();
+        witnessed = new Object();
+        witness = new ExceptionWitness(eventBus, witnessed);
+        Scene scene = new Scene(new ExceptionWitnessResponderButton(eventBus, witnessed, "Test Button Popover"), 800, 600);
+        stage.setScene(scene);
+        stage.show();
+    }
+
+    @Test
+    public void testOnExceptionEvent() throws Exception {
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/components/ExceptionWitnessResponderButtonTest.java]: :+1:

**Comment by ThomasJClark** [comment]
Looks OK but can you do `setDetachable(false);` on the popover?

**Comment by JLLeitschuh** [comment]
I like that you can detach it. Why would we remove that ability?

**Comment by JLLeitschuh** [comment]
They will automatically disappear once resolved even when detached.

**Comment by ThomasJClark** [comment]
I know, but I think you always want it to stay next to the step that it applies to

**Comment by JLLeitschuh** [comment]
I disagree. Being able to move it to see the step better could be helpful.
It won't start detached. 
Also if you do it that way then you can't keep the warning open while you mess with the step.

**Pull request merged**